### PR TITLE
💦 building widget state

### PR DESCRIPTION
**Pull request by stevejpurves**
Aims:
* extend notebook processing to lift widget state from notebook metadata
* notebook metadata can be big and should be out-of-band

Todo:
- [x] first pass
- [x] don't duplicate metadata (e.g. kernel specs)
- [x] decide on pattern for out-of-band

### CODE REVIEW

File: packages/myst-cli/src/process/mdast.ts [hunk 1]
```diff
@@ -39,7 +40,7 @@ import { VFile } from 'vfile';
 import { processPageFrontmatter, updateFileInfoFromFrontmatter } from '../frontmatter.js';
 import { selectors } from '../store/index.js';
 import type { ISession } from '../session/types.js';
-import { castSession } from '../session/cache.js';
+import { castSession, writeToCache } from '../session/cache.js';
```
Comment (fwkoch) [inline packages/myst-cli/src/process/mdast.ts:43]: Just following up on the `lint` suggestion.

```suggestion
import { castSession } from '../session/cache.js';
```

File: packages/myst-cli/src/process/mdast.ts [hunk 2]
```diff
@@ -256,7 +258,9 @@ export async function transformMdast(
     frontmatter,
     mdast,
     references,
-  };
+    widgets,
+    something: 'else',
```
Comment (fwkoch) [inline packages/myst-cli/src/process/mdast.ts:262]: 🧹

```suggestion
```

File: packages/myst-cli/src/process/notebook.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ import type {
   ICell,
   IMimeBundle,
   INotebookContent,
+  INotebookMetadata,
```
Comment (fwkoch) [inline packages/myst-cli/src/process/notebook.ts:10]: 🧹

```suggestion
```

File: packages/myst-cli/src/process/site.ts [hunk 4]
```diff
@@ -376,7 +377,10 @@ export async function fastProcessFile(
     pageReferenceStates,
     extraLinkTransformers,
   });
-  const { mdast, frontmatter } = castSession(session).$getMdast(file)?.post ?? {};
+  const { mdast, frontmatter, widgets } = castSession(session).$getMdast(file)?.post ?? {};
+  console.log('fast process');
+  if (widgets) console.log('got widgets!!');
```
Comment (fwkoch) [inline packages/myst-cli/src/process/site.ts:382]: 🧹

```suggestion
```

File: packages/myst-cli/src/process/site.ts [hunk 5]
```diff
@@ -471,7 +475,10 @@ export async function processProject(
   if (writeFiles) {
     await Promise.all(
       pages.map(async (page) => {
-        const { mdast, frontmatter } = castSession(session).$getMdast(page.file)?.post ?? {};
+        const { mdast, frontmatter, widgets } =
+          castSession(session).$getMdast(page.file)?.post ?? {};
+        console.log('process project - full process');
+        if (widgets) console.log('got widgets!!');
```
Comment (fwkoch) [inline packages/myst-cli/src/process/site.ts:481]: 🧹

```suggestion
```

File: packages/myst-cli/src/process/mdast.ts [hunk 6]
```diff
@@ -334,6 +338,41 @@ export async function postProcessMdast(
   if (checkLinks) await checkLinksTransform(session, file, mdast);
 }
 
+/**
+ * Write widget state to file in public folder
+ */
+export function transformWidgetsToFile(
```
Comment (fwkoch) [inline packages/myst-cli/src/process/mdast.ts:344]: Steve and I chatted about this - there is some duplication between this and the notebook `outputs` caching. These also follow a slightly different pattern than other on-disk caching that is saved to the `/cache` folder. Definitely potential to normalize and deduplicate code... but for now, I think this standalone function is fine.

**Review by fwkoch — APPROVED** [review]
Just a few things leftover from testing to tidy up, but overall this looks good. Serializing widget state will open up the door to some cool stuff!!

File: packages/myst-cli/src/process/mdast.ts [hunk 7]
```diff
@@ -254,7 +255,9 @@ export async function transformMdast(
     frontmatter,
     mdast,
     references,
-  };
+    widgets,
+    something: 'else',
```
Comment (rowanc1) [inline packages/myst-cli/src/process/mdast.ts:259]: Needs a final pass!

File: packages/myst-cli/src/process/site.ts [hunk 8]
```diff
@@ -348,6 +349,7 @@ export async function fastProcessFile(
     extraLinkTransformers,
   });
   const { mdast, frontmatter } = castSession(session).$getMdast(file)?.post ?? {};
+  console.log('fast process');
```
Comment (rowanc1) [inline packages/myst-cli/src/process/site.ts:352]:
```suggestion
```

File: packages/myst-cli/src/process/site.ts [hunk 9]
```diff
@@ -439,6 +441,7 @@ export async function processProject(
     ),
   );
   // Write all pages
+  console.log('full process');
```
Comment (rowanc1) [inline packages/myst-cli/src/process/site.ts:444]:
```suggestion
```

**Review by rowanc1 — APPROVED** [review]

**Pull request merged**